### ISSUE TITLE
Automatic schedule updates

### ISSUE THREAD

**Issue by derek-adair** [derek-adair/nflgame#1]
**I would like to be lazy and not have to run update_schedule.py --year every time the schedule changes, a game is added or a new season starts.**

This is one of most common issues submited to burnsushi/nflgame, "Why cant i see this years games?"I need to drudge up the tickets, but at some point I brought this up and there were some surface level discussion about how this may work. Just getting the ball rolling on this because i feel like it could have a domino effect on a lot of this project.

Any thoughts on this?  Does anyone have a clever solution that I am missing?

**Comment by derek-adair** [comment]
>The schedule could change!

Yup! This is the main reason for me.  Last year there was a cancellation which was a bit of a pain to handle w/ my personal setup.

**Comment by derek-adair** [comment]
>.... would have to be updated every week

Oh hey good catch, i think you ca probably remove the week paramater and get the entire schedule.

**Comment by derek-adair** [comment]
Another potentially troubling issue is that the data returned could be different and/or incomplete.

**Comment by ochawkeye** [comment]
The schedule definitely changes throughout the year, even without a hurricane like the one that that moved a game by 10 weeks last season. 
Most of the second half of the season Sunday night games are "flex" games.  They can and do shuffle around a few weeks before they occur.

In most cases, a user really only needs to be aware of a schedule change the week that the games are actually happening.  That is part of the reason that `update_schedule.py` by default only updates the current week (but can be overridden with the `--year` flag).  Since `nflgame.live.current_year_and_week()` is no longer functioning, this piece is [broken](https://github.com/BurntSushi/nflgame/blob/master/nflgame/live.py#L355-L371).  But if there were an alternate way to inform `nflgame` of the current week that functionality could be returned.

Could be as simple as a calendar until we find something better
```python pseudo code
if date between September 4th 2018 and September 10th 2018:
    _cur_year = 2018
    _cur_week = 1
```
Not very efficient long-term, but would be an easy short term fix.

I do notice that NFL.com has a new score strip on their pages.  A whole lotta JavaScript running it, but maybe someone with more network sniffing acumen could locate a concise source of that info (if it exists)?

Cross-posting with https://github.com/BurntSushi/nflgame/issues/363

**Comment by derek-adair** [comment]
I can take a look i've been thinking of various ways we can make this project better and  one of my big complaints is the way in which the code interacts with nfl.com.  A lot of this, as it turns out, is because nfl.com has no clear way to interact with their data.

**Comment by ochawkeye** [comment]
> A lot of this, as it turns out, is because nfl.com has no clear way to interact with their data.

That's...that's kind of the entire point.  If the NFL had an API this library wouldn't be necessary.  This library is not sanctioned by NFL.com.

Maybe this is just semantics, but to be clear, *none* of the data is manipulated, queried, etc. directly from NFL.com.  That would lead to thousands of calls to NFL.com and most certainly an eventual - if not nearly immediate - cease and desist request from them.

The data all comes from the .json files that run the NFL.com Game Center pages.  Those files are pulled [local ](https://github.com/BurntSushi/nflgame/tree/master/nflgame/gamecenter-json), parsed, and read into memory (or stored in a database in the case of `nfldb`) where they can manipulated, queried, etc.

I do not agree that a complete understanding of the code is necessary to work with the library.

**Comment by mickeelm** [comment]
Maybe a separate issue for the new scorestrip-JS? Otherwise regarding the latest discussion - what @ochawkeye said. 

Regarding the scheduling. If we begin with

> But if there were an alternate way to inform nflgame of the current week that functionality could be returned.

I think that it could be pretty robustly fixed just using the scheduling rules (the regular season starts the weekend after Labor day), which are available on Wikipedia. I've used this (comparing today's date with the scheduling rules to rule out which week we're at) in my own application and it has worked for me. I'll just have to translate it into Python. I'll get on it.

Also, @derek-adair, I need some clarification on this:

> I would like to be lazy and not have to run update_schedule.py --year every time the schedule changes, a game is added or a new season starts.

Like, do you never want to run the script at all or not with the `--year` parameter specifically?

I don't mind running the script, but I'd like it to behave like this

- Using the season scheduling logic, determine where we are in a season (or after/before a season).
- Check the last post of `schedule.json` to see if we're up to date. If the script is run in PRE3 for instance and the last game in the file is POST4 (Super Bowl), fetch PRE1, PRE2, PRE3 and possibly x weeks ahead (PRE4, REG1...). 
- parameters like `--year` still do what they did before.

**Comment by derek-adair** [comment]
>I do not agree that a complete understanding of the code is necessary to work with the library.

I did not mean this.  It is quite user friendly to the end user.  Not so much if you are seeking to modify and/or debug things.  Just some minor organization things, really nothing major.  Anyways, my comment was probably out of place anyways and deserves a separate issue for addressing the scorestrip being depreicated.

Also your description is very good and desrves to be in the wiki or somewhere because it would have saved me some time :D

**Comment by mickeelm** [comment]
Still kinda hijacking this thread, sorry for that but I'm digging through the code a bit to try and understand all that's going on and then I noticed two things that I'd thought I'd bring up:

1. I haven't tried to sniff traffic in order to find the feed that is feeding the score bar on top of NFL.com. The data used to come from http://www.nfl.com/liveupdate/scorestrip/scorestrip.json though (maybe you guys knew about this already) which now seems to be deprecated. I don't know if this new score bar has any more data from a new feed but I doubt it, graphically it kinda looks the same. *Anyhow*, for us this shouldn't matter as that now deprecated feed wasn't used in `nflgame` to begin with. From what I can see in the code, `nflgame` uses only the more detailed game feeds, `game.py` has this code:

```python
_jsonf = path.join(path.split(__file__)[0], 'gamecenter-json', '%s.json.gz')
_json_base_url = "http://www.nfl.com/liveupdate/game-center/%s/%s_gtd.json"
```

With that said, it's always nice to know if we have more data feeds that we can use (but I doubt that we'll get any more info than we already get from the full game feed) but the only "pressing" issue (without workarounds) for the season should be the scheduling thing, which I hoped I solved in #8.

2. @derek-adair reading through the comments in `live.py` and [this wiki page](https://github.com/BurntSushi/nflgame/wiki/Getting-live-statistics-at-regular-intervals) I realized that this application could be run in a "forever mode" (at least the live module), which I didn't know about before. And I guess, if `nflgame` was running "in the background" then I 100% agree that the scheduling updates should be automatic. Maybe the whole application should be possible to run in the background, as a service (systemd etc). Or for that sake, in a python interpreter if you want to keep it in the foreground:

```python
>>> import nflgame
>>> nflgame.run()

```

something like that. Like, that "run mode" would update the schedule, game data, player data and keep track of calling NFL.com during games intervals etc. If you think it sounds like a good idea I could write an issue for it.
